File: 2021/day01/day01.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"time"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func getLines(filename string) []string {
	fh, err := os.OpenFile(filename, os.O_RDONLY, os.ModePerm)
	check(err)
	defer fh.Close()

	lines := make([]string, 0)
	sc := bufio.NewScanner(fh)
	for sc.Scan() {
		line := sc.Text()
		lines = append(lines, line)
	}
	check(sc.Err())
	return lines
}

func part1(lines []string) int {
	cnt := 0
	cur, err := strconv.Atoi(lines[0])
	check(err)
	for _, x := range lines {
		a, err := strconv.Atoi(x)
		check(err)
		if a > cur {
			cnt = cnt + 1
		}
		cur = a
	}
	return cnt
}

func part2(lines []string) int {
	cnt := 0
	cur := 0
	for i, x := range lines {
		if i > 1 && i < len(lines)-1 {
			a, err := strconv.Atoi(x)
			check(err)
			b, err := strconv.Atoi(lines[i-1])
			check(err)
			c, err := strconv.Atoi(lines[i-2])
			check(err)
			sum := a + b + c
			if sum > cur {
				cnt = cnt + 1
			}
			cur = sum
		}
	}
	return cnt
}

func main() {
	timeStart := time.Now()
	argv := os.Args
	if len(argv) < 2 {
		fmt.Printf("Usage: %v /path/to/file\n", argv[0])
		return
	}
	lines := getLines(argv[1])
	fmt.Printf("# Inputs  %d\n", len(lines))
	elapsed := time.Since(timeStart)
	fmt.Printf("# Parsing %s\n", elapsed)
	timeStart1 := time.Now()
	p1 := part1(lines)
	elapsed1 := time.Since(timeStart1)
	fmt.Printf("# Part1   %s\n", elapsed1)
	timeStart2 := time.Now()
	p2 := part2(lines)
	elapsed2 := time.Since(timeStart2)
	fmt.Printf("# Part2   %s\n", elapsed2)
	fmt.Printf("# Total   %s\n", time.Since(timeStart))

	fmt.Printf("Part 1: %d\n", p1)
	fmt.Printf("Part 2: %d\n", p2)
}

File: 2025/day01.py
import sys
import timeit

part1 = True
if len(sys.argv) > 2 and sys.argv[2]:
    part1 = False

lines = []
with open(sys.argv[1], 'r') as fh:
    lines = fh.readlines()
    lines = list(map(lambda s: s.strip(), lines))

cur = 50
rv1 = 0
rv2 = 0

for line in lines:
    num = int(line[1:])
    if line[0] == "R":
        cur = cur + num
        while cur > 99:
            cur = cur - 100
    else:
        cur = cur - num
        while cur < 0:
            cur = cur + 100
    if cur == 0:
        rv1 = rv1 + 1

cur = 50

for line in lines:
    num = int(line[1:])
    if line[0] == "R":
        for i in range(0, num):
          cur += 1
          if cur == 100:
            cur = 0
          if cur == 0:
            rv2 += 1
            #print(rv2)
    elif line[0] == "L":
        for i in range(0, num):
          cur -= 1
          if cur == -100:
            cur = 0
          if cur == 0:
            rv2 += 1
    #print("    ...",last)


if part1:
    print(rv1)
else:
    print(rv2)
